jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IO.java

```diff
@@ -224,36 +224,110 @@ public static void copyDir(File from, File to) throws IOException
+        }
+    }
+
+    /**
+     * Copy the contents of a source directory to destination directory.
+     *
+     * @param srcDir the source directory
+     * @param destDir the destination directory
+     * @param copyOptions the options to use on the {@link Files#copy(Path, Path, CopyOption...)} commands.
+     * @throws IOException if unable to copy the file
+     */
     public static void copyDir(Path srcDir, Path destDir, CopyOption... copyOptions) throws IOException
     {
         if (!Files.isDirectory(Objects.requireNonNull(srcDir)))
             throw new IllegalArgumentException("Source is not a directory: " + srcDir);
         if (!Files.isDirectory(Objects.requireNonNull(destDir)))
             throw new IllegalArgumentException("Dest is not a directory: " + destDir);
 
-        try (Stream<Path> sourceStream = Files.walk(srcDir, 20))
+        try (Stream<Path> sourceStream = Files.walk(srcDir))
         {
             Iterator<Path> iterFiles = sourceStream
                 .filter(Files::isRegularFile)
                 .iterator();
             while (iterFiles.hasNext())
             {
                 Path sourceFile = iterFiles.next();
-                URI relativeSrc = srcDir.toUri().relativize(sourceFile.toUri());
-                Path destFile = destDir.resolve(relativeSrc.toASCIIString());
+                Path relative = srcDir.relativize(sourceFile);
+                Path destFile = resolvePath(destDir, relative);
                 if (!Files.exists(destFile.getParent()))
                     Files.createDirectories(destFile.getParent());
                 Files.copy(sourceFile, destFile, copyOptions);
             }
         }
     }
 
+    /**
+     * Perform a resolve of a path against a relative path in a
+     * way that is smart about {@link java.nio.file.FileSystem} differences.
+     *
+     * @param basePath the base Path
+     * @param relative the Path to resolve against base Path
+     * @return the new Path object relative to the base Path
+     */
+    public static Path resolvePath(Path basePath, Path relative)
+    {
+        if (basePath.getFileSystem().equals(relative.getFileSystem()))
+        {
+            return basePath.resolve(relative);
+        }
+        else
+        {
+            for (Path segment : relative)
+                basePath = basePath.resolve(segment.toString());
+            return basePath;
+        }
+    }
+
+    public static void copyFile(Path srcFile, Path destFile) throws IOException
+    {
+        if (!Files.isRegularFile(Objects.requireNonNull(srcFile)))
+            throw new IllegalArgumentException("Source is not a file: " + srcFile);
+        Objects.requireNonNull(destFile);
+
+        if (!Files.exists(destFile) && srcFile.getFileSystem().equals(destFile.getFileSystem()))
+        {
+            Files.copy(srcFile, destFile, StandardCopyOption.COPY_ATTRIBUTES);
+        }
+        else
+        {
+            try (InputStream in = Files.newInputStream(srcFile);
```

Comment:
Can't `Files.copy()` work in all cases with appropriate flags?

Reply:
No, it causes problems in the foreign target scenario, as that will attempt to delete before copy.
We want to avoid that delete.